### DISCUSSION TITLE
Assignable identifiers for Polybar instances that can be used for IPC

### DISCUSSION THREAD

**Discussion by blaxpot** [polybar/polybar#2912] [Ideas]
It would be great if I could assign each instance of Polybar an identifier that I could use with `polybar-msg`. This is currently possible using PIDs (e.g. `polybar-msg -p 1234 cmd show`), but that's not ideal since the PIDs are ephemeral. I'd love to be able to assign each instance a name that I could use to send it commands instead.

Here's what I would use this for and my current solution, in case there's already a way to do what I want that I'm not seeing:

I want to have a single bar that is hidden by default and is shown on the currently focused monitor only when I hold down a key.

My first instinct was to use a single instance of Polybar and move it around somehow as needed. The only way I can find to do that was by reloading or restarting it. This wasn't ideal, however, since restarting Polybar or reloading the config. seemed to be a bit resource intensive and caused performance issues (brief slowdown/mouse unresponsiveness when switching monitors) on my (fairly capable) laptop that I found unacceptable.

At the moment what I'm doing is launching an instance of Polybar for each monitor using the same config. and writing a line containing the PID and output name to a file for each of my monitors  in `launch.sh` like so:
```
#!/usr/bin/env sh
truncate -s 0 "${HOME}/.config/polybar/pids"
polybar-msg cmd quit

polybar -m | cut -d : -f 1 |
while IFS= read -r OUTPUT; do
  MONITOR="${OUTPUT}" polybar primary &>/dev/null & disown
  PID=$!
  echo "${PID} ${OUTPUT}" >> "${HOME}/.config/polybar/pids"
done

sleep 1 && polybar-msg cmd hide
```

Then to show a particular instance of Polybar I can do something like this (I'm using i3 but I'm sure you could do something similar with other WMs):
```
#!/usr/bin/env sh
WORKSPACE="$(i3-msg -t get_workspaces | jq -r '.[] | select(.focused) | .name')"
OUTPUT="$(i3-msg -t get_outputs | jq -r ".[] | select(.current_workspace == \"${WORKSPACE}\") | .name")"
PID="$(grep` "${OUTPUT}" "${HOME}/.config/polybar/pids" | cut -d ' ' -f 1)"
polybar-msg -p "${PID}" cmd show
```

The above does what I want it to but it would be a lot more elegant if I could assign IDs to each instance of Polybar when I start them like so:
`MONITOR="${OUTPUT}" polybar -i output1-bottom primary &>/dev/null & disown`

And then show them with:
`polybar-msg -i output1-bottom cmd show`

**Comment by blaxpot** [comment]
Another way do do this, mentioned in https://github.com/polybar/polybar/discussions/2780, would be to use `xdotool` to find a given instance of Polybar, e.g.:
`xdotool search --name polybar-bottom getwindowpid`

I would still prefer to be able to refer to each instance by name rather than having to look up the PID, though.

**Comment by patrick96** [comment]
You can already assign each polybar instance its own name through the [`wm-name`](https://github.com/polybar/polybar/wiki/Configuration#bar-settings) setting. By default, the window is identified by the bar and monitor names, but you could use your own unique identifier.

As you already noted, this can be used to look up the PID through `xdotool`.
I don't really see a good reason to also implement this lookup functionality in `polybar-msg` since `xdotool` can already do it better (also supports regex).

**Comment by blaxpot** [comment]
It's not convenient for a user to have to look up an ephemeral identifier in order to send messages to an instance of an application.

I'm not necessarily suggesting that you implement PID lookup functionality in `polybar-msg`, though. I was thinking more along the lines of adding a field to IPC messages so that they can address specific instances of the application without needing to look anything up.

Ofc, idk the details of how things are implemented at all, so maybe that's not feasible.

**Reply by patrick96** [reply]
A filter might be an interesting idea. Then bars could be addressed by identifier, name, monitor, etc.

But that is still quite far off. The current IPC format is really primitive and doesn't support adding message fields.

**Reply by blaxpot** [reply]
I think it could open up some interesting possibilities for module actions. The use case I'm describing here is very basic in comparison.
